egs2/rats/spk1/local/rats_trial_prep.py

```diff
@@ -1,75 +1,104 @@
+        if spk_id is not None:
+            d_spk2utt.setdefault(spk_id, []).append(line)
 
-        d_spk2utt[spk_id].append(line)
-
-    print("# of spk: {}".format(len(d_spk2utt.keys())))
-    print("# of utt: {}".format(len(l_dev)))
+    print(f"# of spk: {len(d_spk2utt)}")
+    print(f"# of utt: {len(l_dev)}")
 
     used_trials = set()
 
-    def is_unique_trial(class_idx, utt1, utt2):
-        trial = (class_idx, utt1, utt2)
-        reverse_trial = (class_idx, utt2, utt1)
-        if trial in used_trials or reverse_trial in used_trials:
-            return False
-        used_trials.add(trial)
-        return True
+    def is_unique_trial(trials):
+        """Check if the trial is unique and add to used trials."""
+        unique_trials = []
+        for trial in trials:
+            class_idx, utt1, utt2 = trial
+            reverse_trial = (class_idx, utt2, utt1)
+            if trial not in used_trials and reverse_trial not in used_trials:
+                used_trials.add(trial)
+                unique_trials.append(trial)
+        return unique_trials
+
+    spk_keys = list(d_spk2utt.keys())
+    dev_set = set(l_dev)
 
     with open(args.out, "w") as f_out:
-        for _ in range(len(d_src2spk.keys()) * 20):
-            spk = np.random.choice(list(d_spk2utt.keys()), size=1)[0]
-            class_idx = np.random.randint(2)
-
-            if class_idx == 0:
-                while True:
-                    utt1, utt2 = np.random.choice(
-                        list(d_spk2utt[spk]), size=2, replace=False
-                    )
-                    if is_unique_trial(class_idx, utt1, utt2):
-                        break
-            elif class_idx == 1:
-                while True:
-                    tmp_lines = list(set(l_dev) - set(d_spk2utt[spk]))
-                    utt1, utt2 = np.random.choice(tmp_lines, size=2, replace=False)
-                    if is_unique_trial(class_idx, utt1, utt2):
-                        break
-
-            f_out.write("{} {} {}\n".format(class_idx, utt1, utt2))
+        # Pre-generate the random speaker and class indices to reduce calls to random
+        n_trials = len(d_src2spk) * 20
+        random_spks = np.random.choice(spk_keys, size=n_trials)
+        class_indices = np.random.randint(0, 2, size=n_trials)
+
+        batch_size = 100  # Process in batches
+
+        for i in tqdm(range(0, n_trials, batch_size)):
+            batch_spks = random_spks[i : i + batch_size]
+            batch_classes = class_indices[i : i + batch_size]
+            batch_trials = []
+
+            for spk, class_idx in zip(batch_spks, batch_classes):
+                if class_idx == 1:
+                    # Same-speaker pair
+                    utt_list = d_spk2utt[spk]
+                    if len(utt_list) >= 2:
+                        utt_pairs = np.random.choice(
+                            utt_list,
+                            size=(min(batch_size, len(utt_list) // 2), 2),
```

Comment:
It's not wrong nor critical, but we could think about this part where it seems to use (almost) all utterances once for target pair.
Could it be better to use some utterances multiple times, paired with different utterances? (an open question, which I don't have answer to)